akka-http-core/src/main/scala/akka/http/Http.scala

```diff
@@ -128,44 +138,182 @@ class HttpExt(config: Config)(implicit system: ActorSystem) extends akka.actor.E
+              throw e
+          }
+
+        case CowboyEntry ⇒ { Thread.`yield`(); hostPool() } // spin while other thread is creating a HostPool for this setup
+        case pool        ⇒ pool // return cached instance
+      }
+    hostPool().clientFlow[T]
+  }
+
+  /**
+   * Creates a new "super connection pool flow", which routes incoming requests to a (cached) host connection pool
+   * depending on their respective effective URI. Note that incoming requests must have either an absolute URI or
+   * a valid `Host` header.
+   *
+   * Since the underlying transport usually comprises more than a single connection the produced flow might generate
+   * responses in an order that doesn't directly match the consumed requests.
+   * For example, if two requests A and B enter the flow in that order the response for B might be produced before the
+   * response for A.
+   * In order to allow for easy response-to-request association the flow takes in a custom, opaque context
+   * object of type ``T`` from the application which is emitted together with the corresponding response.
+   */
+  def superPool[T](options: immutable.Traversable[Inet.SocketOption] = Nil,
+                   settings: Option[HostConnectionPoolSettings] = None,
+                   defaultHeaders: List[HttpHeader] = Nil,
+                   log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Flow[(HttpRequest, T), (Try[HttpResponse], T), Unit] = {
+    val effectiveSettings = HostConnectionPoolSettings(settings)
+    val setup = ConnectionPoolSetup(encrypted = false, options, effectiveSettings, defaultHeaders, log)
+    ConnectionPool.superPool(setup)
+  }
+
+  /**
+   * Fires a single [[HttpRequest]] across a freshly created super pool flow to produce a response future.
+   * Note that incoming requests must have either an absolute URI or a valid `Host` header, otherwise
+   * the future will be completed with an error.
+   */
+  def singleRequest(request: HttpRequest,
+                    options: immutable.Traversable[Inet.SocketOption] = Nil,
+                    settings: Option[HostConnectionPoolSettings] = None,
+                    defaultHeaders: List[HttpHeader] = Nil,
+                    log: LoggingAdapter = system.log)(implicit fm: FlowMaterializer): Future[HttpResponse] = {
+    val effectiveSettings = HostConnectionPoolSettings(settings)
+    val setup = ConnectionPoolSetup(encrypted = false, options, effectiveSettings, defaultHeaders, log)
+    Source.single(request -> ())
+      .via(ConnectionPool.superPool(setup))
+      .map(_._1.get)
+      .runWith(Sink.head)
+  }
+
+  /**
+   * Triggers an orderly shutdown of all host connections pools currently maintained by the [[ActorSystem]].
+   */
+  def shutdownAllConnectionPools(): Future[Unit] = {
+    import scala.collection.JavaConverters._
+    import system.dispatcher
+    val hostPools = hostPoolCache.values().asScala
+    Future.sequence(hostPools.map(_.doShutdown())).map(_ ⇒ ())
```

Comment:
if this is supposed to be reliable, then it needs to guard against concurrent or subsequent pool creation as well